bpo-37779 : Add information about the overriding behavior of ConfigParser.read (#15177)
python/cpython: Doc/library/configparser.rst

sections [1]_.  Note also that keys in sections are
case-insensitive and stored in lowercase [1]_.

If it is necessary to read several configurations, with each one having more
priority than the previous one, an instance of :class:`ConfigParser` can be

> used to override previously defined properties and retain existing ones.

Reviewer: One problem I see with this is that the first few clauses of this sentence seem to be introducing a tutorial, i.e. "If you want to do X, you can achieve it with Y", but then it turns out that you are actually *describing the existing behavior* of `ConfigParser` rather than providing a recipe or something.

I would recommend rewriting this as an introduction to the properties of a `ConfigParser`, like so:

> It is also possible to read several configurations into a single ```:class:`ConfigParser` ``` object, in which case the most recent configuration has priority over the previous ones; conflicting keys will be taken from the new configuration and existing keys will be retained.

That's mostly an example, I'm not super happy with that particular wording...

Reviewer 2: @pganssle:
> That's mostly an example, I'm not super happy with that particular wording...

What do you think of this wording?

> It is possible to read several configurations into a single :class:`ConfigParser`, where the most recently added configuration has the highest priority. Any conflicting keys are taken from the more recent configuration while the previously existing keys are retained.

Reviewer: :shipit: That looks great!

Author: @pganssle I pushed @aeros167 's version to the repository.

Author: I have made the requested changes; please review again

.. doctest::

   >>> config = configparser.ConfigParser()
   >>> config.read('example.ini')
   ['example.ini']
   >>> config['topsecret.server.com']['Port']
   '50022'
   >>> config.read_string("[topsecret.server.com]\nPort=48484")
   >>> config['topsecret.server.com']['Port']
   '48484'
   >>> config.read_dict({"topsecret.server.com": {"Port": 21212}})
   >>> config['topsecret.server.com']['Port']
   '21212'
   >>> config['topsecret.server.com']['ForwardX11']
   'no'

This behaviour is equivalent to a :meth:`ConfigParser.read` call with several

> files passed to *filenames* parameter.

Reviewer: ```suggestion
files passed to the *filenames* parameter.
```

> files passed to ``filenames`` parameter.

Reviewer 2: ```suggestion
files passed to *filenames* parameter.
```

case-insensitive and stored in lowercase [1]_.

> 

Reviewer 3: nit: single line here.

Author: This line is not modified by this PR. So I wonder if it's a good place to fix it.

What do you think about it?

Reviewer 3: What I know is there should be a single line before the next paragraph in a section. not two.

Author: I think it's ok now.

> If we need to read several configurations, each one having more priority than the previous one, we can use the same :class:`ConfigParser` instance to override previous defined data and keep not redefined data.

Reviewer 3: Am not a native English speaker but this "keep not redefined data" phrase is confusing at least to me. Not sure how you can improve it.

Reviewer 2: Thanks for the contribution @sblondon and suggestion @nanjekyejoannah.

A simple grammatical improvement would be to change "keep not redefined data" to "not keep redefined data". However, I would suggest a few other changes to this part as well. This retains the same meaning, but significantly improves the phrasing:
```suggestion
If it is necessary to read several configurations, with each one having more 
priority than the previous one, an instance of :class:`ConfigParser` can be 
used to override previously defined properties and retain existing ones.
```
reST and Sphinx specifications (this applies to the other sections as well):
1) The maximum number of characters per line is 80 characters. 

Significant phrasing changes:
1) In technical writing, it is preferable to avoid using second person pronouns, such as "we" when possible.
2) "data" -> "properties". This term is more of an accurate descriptor than "data". An alternative term would be "attributes".
3) "keep not redefined data" -> "retain existing ones". This removes redundancy, "ones" refers to the previously mentioned properties, which implies "retain existing properties". 

Edit: Removed part about 3 spaces, see my comment below.

> If we need to read several configurations, each one having more priority than the previous one, we can use the same :class:`ConfigParser` instance to override previous defined data and keep not redefined data.

Reviewer 3: I would just say  "use" instead of "we can use".

Author: I reused aeros167's version so do you think this change is needed?
The new version is `...an instance of :class:ConfigParser can be
used to override...`.

Reviewer 3: No problem.

.. doctest::

   >>> config = configparser.ConfigParser()
   >>> config.read('example.ini')
   ['example.ini']
   >>> config['topsecret.server.com']['Port']
   '50022'
   >>> config.read_string("[topsecret.server.com]\nPort=48484")
   >>> config['topsecret.server.com']['Port']
   '48484'
   >>> config.read_dict({"topsecret.server.com": {"Port": 21212}})
   >>> config['topsecret.server.com']['Port']
   '21212'
   >>> config['topsecret.server.com']['ForwardX11']
   'no'

This behaviour is equivalent to a :meth:`ConfigParser.read` call with several files passed to ``filenames`` parameter.

> 

Reviewer 3: nit: Add one more line before the new section.

Author: Fixed
